compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs

```diff
@@ -2202,8 +2217,138 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
         }
     }
 
-    /// Literals and const generic parameters are eagerly converted to a constant, everything else
-    /// becomes `Unevaluated`.
+    #[instrument(level = "debug", skip(self), ret)]
+    pub fn lower_const_assoc_path(
```

Comment:
We should definitely consolidate this and `lower_assoc_path` into one method that handles 99% of the logic since the only real behavioural differences are the `AssocKind` arguments we hard code and that we construct a `Ty` or `Const` at the very end depending on which method it is

Reply:
Done. Not sure if the new code is the greatest ever, but seems to work.

Reply:
Yeah it's much better now :D left a couple nits as a proper review comment but this is much nicer and definitely solved the core problem of duplicating the tricky logic